Gömülü Sistemler dersinde bitwise işlemleri çok sık kullanacağız:  
\& : VE  
| : VEYA  
~ : DEĞİL  
^ : XOR  
<< : Sola kaydırma  
\>> : Sağa kaydırma  

In [2]:
unsigned int a=12;
unsigned int b=10;

Yukarıdaki atama işlemleri sonucu bellekte 2 tabanında değer saklanır. Bu sayıyı 10 tabanında veya 16 tabanında gösterebiliriz.

In [3]:
a
b

Dec: 12 Hex: C Bin: 00001100
Dec: 10 Hex: A Bin: 00001010


VE işleminde her iki bit de 1 ise sonuç 1 olur diğer durumlarda sonuç 0 olur.

In [12]:
a
b
a&b

00001100
00001010
00001000


VEYA işleminde bitlerden 1 tanesinin  1 olması sonucu 1 yapmak için yeterlidir.

In [14]:
a
b
a|b

00001100
00001010
00001110


XOR işlemi (mod 2)'de toplama işlemi gibi düşünülebilir.

In [15]:
a
b
a^b

00001100
00001010
00000110


DEĞİL işleminde bit 0 ise 1, 1 ise 0 yapılır.

In [16]:
a
~a

00001100
11110011


Sağa kaydırma operatörü: >> 

In [18]:
a
a>>1
a>>2

00001100
00000110
00000011


Sola kaydırma operatörü: <<

In [22]:
a
a<<1
a<<2

00001100
00011000
00110000


**Set Bit** İşlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitini 1 yapma işlemidir.

a değişkeninin n=10. bitini set edelim.

In [4]:
int n;

In [5]:
n=10;
a
a | (1<<n)

a: 0000000000001100
   0000010000001100


**Clear Bit** İşlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitini 0 yapma işlemidir.

a değişkeninin n=3. bitini temizleyelim.


In [32]:
n=3;
a
a & ~(1<<n)

a: 0000000000001100
   0000000000000100


Test Bit işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. bitinin 1 veya  0 olduğunu kontrol etme işlemidir.

In [36]:
n=3;
a
(a >> n) & 1

a: 0000000000001100
   1


In [37]:
n=4;
a
(a >> n) & 1

a: 0000000000001100
   0


**Toggle Bit** işlemi, bir değişkeninin en sağdaki biti 0. biti olmak üzere n. biti 1 ise 0 yapma 0 ise 1 yapma işlemidir.

In [9]:
n=2;
a
a ^ (1<<n)
n=4;
a ^ (1<<n)

a: 0000000000001100
   0000000000001000
   0000000000011100


Clear Range: bir bölgenin temizlenmesi (0 yapılması) işlemidir.   
Örneğin 7. bitten itibaren 5 bitlik bir bölge 0 yapılacaksa:

In [ ]:
unsigned int cr;

In [32]:
cr=0x1F;

In [33]:
a=7917;
n=7;
a
a & ~(cr<<n)

a: 0001111011101101
   0001000001101101


Set Range: bir bölgeye bir verinin yazılması işlemidir.  
Bu işlemi yapabilmek için önce o bölgenin temizlenmesi gerekmektedir.  
Temzilenme işleminden sonra yazılmak istenen veri VEYA işlemi ile belirlenen bölgeye yazılır.

9\. bitinden itibaren 1010 verisini yazalım:

In [35]:
int d;

In [41]:
a=7917;
cr=0xF;
d=0xA;
n=9;
a
(a & ~(cr<<n)) | (d << n))

a: 0001111011101101
   0001010011101101
